crypto/rand/rand_lib.c

```diff
@@ -174,8 +174,6 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
                                    prediction_resistance,
                                    (unsigned char *)&drbg, sizeof(drbg)) != 0)
                 bytes = bytes_needed;
-            drbg->reseed_next_counter
-                = tsan_load(&drbg->parent->reseed_prop_counter);
```

Comment:
This was supposed to synchronize the reseed counter with the parent.
Why did this not work?

Reply:
I'm not saying that your fix didn't work, but it came at the price of an unnecessary complification of thje code.

For example, if you look at RAND_DRBG_reseed(), the secondary (i.e. public) DRBG [increments the reseed_next_counter unconditionally](https://github.com/openssl/openssl/pull/12759/files#diff-9181ac017a6177a5f2619f65c9b7a346L432-L437) without checking whether it is chained or not. Then it calls `drbg->get_entropy()`, where the reseed_next_counter happens to be reset to the parent's counter (by the statement you pointed out), and finally, [the reseed_prop_counter gets updated](https://github.com/openssl/openssl/pull/12759/files#diff-9181ac017a6177a5f2619f65c9b7a346L456).

If it it had been your change only, that might have been tolerable. But as a consequence of the increased complexity and some unnecessary renamings and re-renamings, the simple concept has deteriorated into [something completely different](https://github.com/openssl/openssl/commit/714a1bb380ddb2bf7538f6a61f47ac87200e3e06)  on the master branch: the secondary and primary DRBGs maintain counters which are incremented independently and the secondary DRBG takes the lock to read the primary DRBG's counter.

My primary goal is to repair the implementation on master, but I also want to restore the simplicity of the original approach and the consistency between the two branches. After all, all what was missing from my original solution was to [convert some increments and assignments to atomic operations](https://github.com/openssl/openssl/pull/12759/commits/d86bb47aab98dfdb7c222bcc84148991fbc0f42a).